### ISSUE TITLE
Atlas gives "SystemStackError: stack level too deep" for a child converter whose parent converters have hard-coded zero demand

### ISSUE THREAD

**Issue by RobTerwel** [quintel/etsource#982]
Atlas returns "SystemStackError: stack level too deep" (often an indication of recursion) for a child converter whose parent converters have their demand preset to 0. Here the preset demands (from the EB) are 

- `transport_final_demand_for_road_lng` = non-zero
- `transport_final_demand_for_shipping_lng`= 0
- `transport_final_demand_for_road_lng`= 0
- `transport_final_demand_for shipping_lng` = 0 
- and their sums, ` transport_final_demand_lng` and `transport_final_demand_bio_lng`

This yields

>   Calculating the ETSource datasets calculates the nl dataset
     Failure/Error: Unable to find matching line from backtrace
     SystemStackError:
       stack level too deep
     # /Users/robterwel/.rbenv/versions/2.1.2/lib/ruby/gems/2.1.0/bundler/gems/refinery-58c1138fd1a7/lib/refinery/edge.rb:19

I looked at the Ruby file mentioned here and line 19 is part of a definition function and says:

> get(:demand)

'Overdetermining' the system, by setting more shares and preset demands than necessary, but which are in agreement with the originally preset demands, does not help. See the image of the initial graph below.

![screen shot 2015-10-21 at 09 45 07](https://cloud.githubusercontent.com/assets/14106387/10632137/ae5c3376-77e2-11e5-91ce-4d8e17f75966.png)

E.g., in this case both `transport_final_demand_for_shipping_lng` and `transport_final_demand_for_shipping_bio_lng` are zero, meaning that `transport_shipping_mixer_lng` should also become zero. At first it seemed to me that Atlas probably cannot determine the shares of two converging flows which are zero, but presetting all these shares did not solve the issue either.  

However, when a very small or any non-zero demand is set for `transport_final_demand_for_shipping_bio_lng` or `transport_final_demand_for_shipping_lng`, everything passes and works. An interesting implication of this is that the problem probably lies *down* the road, with the child converter `transport_shipping_mixer_lng`or beyond, since there is no parent share for `transport_final_demand_bio_lng` and all demand is zero for bio-lng.  See below for the initial graph.

![screen shot 2015-10-21 at 12 17 23](https://cloud.githubusercontent.com/assets/14106387/10633907/51310918-77ee-11e5-9d20-382426bad1cf.png)

In this case I added a small value to the 0 the energy_balance gave for `transport_final_demand_for_shipping_lng`. In https://github.com/quintel/etsource/commit/e00d115b22ca362c1bd972dfcaf39fc6b6b7a3ec , I added a small value to both that demand and `transport_final_demand_for_road_bio_lng` to make Semaphore pass for all countries, since some have zero demand for all aforementioned converters according to the EB. The challenge now is to find a way to initialise the graph without errors and without these added values. 

@antw, would you have any clue about how I could make this work or why it does not work yet?

**Comment by antw** [comment]
This reminds me of quintel/atlas#46 when a share value of zero resulted in an incalculable graph (but not a stack error). I tried e00d115b, and the commit prior to that, but couldn't reproduce the error; could you point me to the commit where this error happens?

**Comment by RobTerwel** [comment]
I just pushed a new commit https://github.com/quintel/etsource/commit/870e01dd883da2acc33a00ac3138d73ca3053069 with the correct (hard-coded 0) values, so it should now reproduce the error (and indeed it does, because it fails the checks). 

In the previous commit  e00d115 I put different (incorrect) values to make it pass.

**Comment by antw** [comment]
I see what the problem is now. `transport_shipping_mixer_lng` has two inputs: `lng` and `bio_lng`. Since the input of both is set to zero, the calculation can't figure out what the share of those carriers should be.

For example:

* if 7.5 `lng` and 2.5 `bio_lng` goes in, we can determine that the share of inputs are 75% `lng` and 25% `bio_lng`
* if 7.5 `lng` and 0.0 `bio_lng` goes in, we know the share is 100% `lng`.

However, if there is zero `lng` *and* `bio_lng` entering the converter, the calculation cannot figure out what the share of each is.

--

Do we know what the share of these carriers are? If so, adding the following two lines to [the converter ".ad" file](https://github.com/quintel/etsource/blob/870e01dd883da2acc33a00ac3138d73ca3053069/nodes/transport/transport_shipping_mixer_lng.converter.ad) will do the trick:

```
- input.lng = 0.1      # <- swap with real value
- input.bio_lng = 0.9  # <- swap with real value
```

[Here's a real-world example on the cokesoven converter](https://github.com/quintel/etsource/blob/870e01dd883da2acc33a00ac3138d73ca3053069/nodes/energy/energy_cokesoven_consumption_coal_gas.converter.ad#L8-L10).

**Comment by RobTerwel** [comment]
Thank you very much, @antw ! You've been a great help.

Initially I thought setting child shares for the mixers might do the trick, but that didn't work; fortunately, your solution of setting inputs solves all my trouble! For the time being, I defined these converter inputs in terms of shares and EB values and that works perfectly. Now I'll find a way to generate csv files which do the job and simply substitute the temporary calculations in the ad file with the csv shares.

Closing.
